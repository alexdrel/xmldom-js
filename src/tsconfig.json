{
  "compilerOptions": {
    "target": "es5",
    "module": "es6",
    "strict": true,
    "esModuleInterop": true,
    "outDir": "../dist",
    "declaration": true
  }
}
