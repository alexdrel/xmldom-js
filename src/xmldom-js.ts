export const string = (a: string) => a;

type Parser = (s: string) => any;
class ContentNode {
  constructor(public parser: Parser) { }
}
export function content(p: Parser = string) {
  return new ContentNode(p);
}
class AttributeNode {
  constructor(public parser: Parser) { }
}
export function attribute(p: Parser = string) {
  return new AttributeNode(p);
}

export type NameResolver = (jn: string, e: Element | Attr) => boolean;

export const ignoreNamespace: NameResolver = (jn, e) => e.localName == jn;

export const prefixNamespace: NameResolver = (jn, e) => e.nodeName == jn;

export function namespaces(def: string, nsmap: { [prefix: string]: string }): NameResolver {
  return (jn, e) => {
    let [p, n] = jn.split(":");
    if (n == undefined) {
      return e.localName == jn && e.namespaceURI == (e.nodeType !== e.ATTRIBUTE_NODE ? def : null);
    }
    return e.localName == n && e.namespaceURI == nsmap[p];
  }
}

function findMap<T extends Element | Attr>(list: ArrayLike<T>, name: string, ns: NameResolver, func: (e: T) => void, once = true) {
  for (let i = 0; i < list.length; i++) {
    if (ns(name, list[i])) {
      func(list[i]);
      if (once) {
        return;
      }
    }
  }
}

export function readXML(xml: Element | Document, schema: any, ns: NameResolver = ignoreNamespace): any {
  const json: any = {};
  const children = xml.children || /* IE */ (xml.childNodes as any as HTMLCollection);
  Object.keys(schema).forEach(k => {
    const s = schema[k];
    let v: any;
    if (s instanceof ContentNode) {
      if (k === "") {
        v = s.parser(xml.textContent as string);
      } else {
        findMap(children, k, ns, e => v = s.parser(e.textContent as string));
      }
    } else if (s instanceof AttributeNode && xml instanceof Element) {
      findMap(xml.attributes, k, ns, e => v = s.parser(e.textContent as string));
    } else if (Array.isArray(s)) {
      v = [];
      findMap(children, k, ns, e => v.push(readXML(e, s[0], ns)), false);
    } else /* Structured element */ {
      findMap(children, k, ns, e => v = readXML(e, s, ns));
    }
    if (v !== undefined) {
      json[k] = v;
    }
  });
  return json;
}
