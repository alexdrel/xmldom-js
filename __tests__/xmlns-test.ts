import { content, string, readXML, attribute, ignoreNamespace, prefixNamespace, namespaces } from "../src/xmldom-js";

const xmlString = `<?xml version='1.0' encoding='UTF-8'?>
    <root xmlns="//ns1" xmlns:ns3="//ns3">
      <empty/>
      <ns2:items xmlns:ns2="//ns2">
        <item id="2" ns2:val="3">
          <ns2:name>Item Name</ns2:name>
        </item>
      </ns2:items>
      <name ns3:a="1">Name</name>
    </root>`;

const xml = new DOMParser().parseFromString(xmlString, "text/xml");

describe("local namespaced xml", () => {
  const json = readXML(xml,
    {
      root: {
        name: {
          "": content(string),
          "ns3:a": attribute(),
          na: attribute()
        },
        empty: content(),
        missing: content(),
        "ns2:items": {
          item: [{
            "ns2:name": content(string),
            "ns2:val": attribute(x => +x),
            id: attribute(x => +x),
            na: attribute()
          }]
        }
      }
    },
    prefixNamespace,
  );

  test("Check the document", () => {
    expect(json.root).toBeDefined();
    expect(json.root.name).toEqual({ "": "Name", "ns3:a": "1" });
    expect(json.root.empty).toBe("");
    expect(json.root.missing).toBe(undefined);
  });

  test("Check the document shot", () => {
    expect(json.root).toMatchSnapshot();
  });
  test("Check items", () => {
    expect(json.root["ns2:items"].item.length).toBe(1);
    const i = json.root["ns2:items"].item[0];
    expect(i).toEqual({ id: 2, "ns2:name": "Item Name", "ns2:val": 3 });

  });
});

describe("ignored namespaced xml", () => {
  const json = readXML(xml,
    {
      root: {
        name: {
          "": content(string),
          a: attribute(),
          na: attribute()
        },
        empty: content(),
        missing: content(),
        items: {
          item: [{
            name: content(string),
            id: attribute(x => +x),
            val: attribute(x => +x),
          }]
        }
      }
    },
    ignoreNamespace
  );

  test("Check the document shot", () => {
    expect(json.root).toMatchSnapshot();
  });
  test("Check items", () => {
    expect(json.root.items.item.length).toBe(1);
    const i = json.root.items.item[0];
    expect(i).toEqual({ id: 2, name: "Item Name", val: 3 });
  });
});

describe("real namespaced xml", () => {
  const json = readXML(xml,
    {
      root: {
        name: {
          "": content(string),
          "ns3:a": attribute(),
          na: attribute()
        },
        empty: content(),
        missing: content(),
        "m:items": {
          item: [{
            "m:name": content(string),
            id: attribute(x => +x),
            "m:val": attribute(x => +x),
            na: attribute()
          }]
        }
      }
    },
    namespaces("//ns1", {
      m: "//ns2",
      ns3: "//ns3",
    }),
  );

  test("Check the document", () => {
    expect(json.root).toBeDefined();
    expect(json.root.name).toEqual({ "": "Name", "ns3:a": "1" });
    expect(json.root.empty).toBe("");
    expect(json.root.missing).toBe(undefined);
  });

  test("Check the document shot", () => {
    expect(json.root).toMatchSnapshot();
  });
  test("Check items", () => {
    expect(json.root["m:items"]).toBeDefined();
    expect(json.root["m:items"].item.length).toBe(1);
    const i = json.root["m:items"].item[0];
    expect(i).toEqual({ id: 2, "m:name": "Item Name", "m:val": 3 });
    expect(i["m:name"]).toBeDefined();
    expect(i.id).toBe(2);
    expect(i.na).toBe(undefined);
  });
});
