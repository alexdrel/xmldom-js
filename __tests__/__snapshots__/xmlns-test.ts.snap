// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ignored namespaced xml Check the document shot 1`] = `
Object {
  "empty": "",
  "items": Object {
    "item": Array [
      Object {
        "id": 2,
        "name": "Item Name",
        "val": 3,
      },
    ],
  },
  "name": Object {
    "": "Name",
    "a": "1",
  },
}
`;

exports[`local namespaced xml Check the document shot 1`] = `
Object {
  "empty": "",
  "name": Object {
    "": "Name",
    "ns3:a": "1",
  },
  "ns2:items": Object {
    "item": Array [
      Object {
        "id": 2,
        "ns2:name": "Item Name",
        "ns2:val": 3,
      },
    ],
  },
}
`;

exports[`real namespaced xml Check the document shot 1`] = `
Object {
  "empty": "",
  "m:items": Object {
    "item": Array [
      Object {
        "id": 2,
        "m:name": "Item Name",
        "m:val": 3,
      },
    ],
  },
  "name": Object {
    "": "Name",
    "ns3:a": "1",
  },
}
`;
